internal/typeexpr/get_type_test.go

```diff
@@ -0,0 +1,385 @@
+		},
+		{
+			`list(any)`,
+			true,
+			cty.List(cty.DynamicPseudoType),
+			``,
+		},
+		{
+			`object({})`,
+			false,
+			cty.EmptyObject,
+			``,
+		},
+		{
+			`object({name=string})`,
+			false,
+			cty.Object(map[string]cty.Type{"name": cty.String}),
+			``,
+		},
+		{
+			`object({"name"=string})`,
+			false,
+			cty.EmptyObject,
+			`Object constructor map keys must be attribute names.`,
+		},
+		{
+			`object({name=nope})`,
+			false,
+			cty.Object(map[string]cty.Type{"name": cty.DynamicPseudoType}),
+			`The keyword "nope" is not a valid type specification.`,
+		},
+		{
+			`object()`,
+			false,
+			cty.DynamicPseudoType,
+			`The object type constructor requires one argument specifying the attribute types and values as a map.`,
+		},
+		{
+			`object(string)`,
+			false,
+			cty.DynamicPseudoType,
+			`Object type constructor requires a map whose keys are attribute names and whose values are the corresponding attribute types.`,
+		},
+		{
+			`tuple([])`,
+			false,
+			cty.EmptyTuple,
+			``,
+		},
+		{
+			`tuple([string, bool])`,
+			false,
+			cty.Tuple([]cty.Type{cty.String, cty.Bool}),
+			``,
+		},
+		{
+			`tuple([nope])`,
+			false,
+			cty.Tuple([]cty.Type{cty.DynamicPseudoType}),
+			`The keyword "nope" is not a valid type specification.`,
+		},
+		{
+			`tuple()`,
+			false,
+			cty.DynamicPseudoType,
+			`The tuple type constructor requires one argument specifying the element types as a list.`,
+		},
+		{
+			`tuple(string)`,
+			false,
+			cty.DynamicPseudoType,
+			`Tuple type constructor requires a list of element types.`,
+		},
+		{
+			`shwoop(string)`,
+			false,
+			cty.DynamicPseudoType,
+			`Keyword "shwoop" is not a valid type constructor.`,
+		},
+		{
+			`list("string")`,
+			false,
+			cty.List(cty.DynamicPseudoType),
+			`A type specification is either a primitive type keyword (bool, number, string) or a complex type constructor call, like list(string).`,
+		},
+
+		// More interesting combinations
+		{
+			`list(object({}))`,
+			false,
+			cty.List(cty.EmptyObject),
+			``,
+		},
+		{
+			`list(map(tuple([])))`,
+			false,
+			cty.List(cty.Map(cty.EmptyTuple)),
+			``,
+		},
+
+		// Optional modifier
+		{
+			`object({name=string,age=optional(number)})`,
+			true,
+			cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+				"name": cty.String,
+				"age":  cty.Number,
+			}, []string{"age"}),
+			``,
+		},
+		{
+			`object({name=string,age=optional(number)})`,
+			false,
+			cty.Object(map[string]cty.Type{
+				"name": cty.String,
+				"age":  cty.Number,
+			}),
+			`Optional attribute modifier is only for type constraints, not for exact types.`,
+		},
+		{
+			`optional(string)`,
+			false,
+			cty.DynamicPseudoType,
+			`Keyword "optional" is valid only as a modifier for object type attributes.`,
+		},
+		{
+			`optional`,
+			false,
+			cty.DynamicPseudoType,
+			`The keyword "optional" is not a valid type specification.`,
+		},
+	}
```

Comment:
A couple extra tests I can think of here would be:
* `object({ name = optional(any) })` in constraint mode, which should succeed
* `object({ name = optional(any) })` in exact type mode, which should fail

(The motivation then being to make sure that the handling of `optional` is correctly passing through its caller's flag for whether we're parsing an exact type or a type constraint.)

Reply:
Added in c82e291.

Reply:
Oh heh I forgot when I was thinking about this that the `optional(...)` syntax itself already fails with an error in exact type mode, so although it does still fail it fails for a different reason than I was imagining.

Thanks for adding the two extra tests, either way!